usd/src/usd_parser/USDTransforms.cc

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright (C) 2022 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include "sdf/usd/usd_parser/USDTransforms.hh"
+#include "sdf/usd/usd_parser/USDData.hh"
+
+#include <ignition/math/Pose3.hh>
+#include <ignition/math/Vector3.hh>
+
+namespace sdf
+{
+  // Inline bracke to help doxygen filtering.
+  inline namespace SDF_VERSION_NAMESPACE {
+  //
+  namespace usd
+  {
+
+  void GetAllTransforms(
+    const pxr::UsdPrim &_prim,
+    USDData &_usdData,
+    std::vector<ignition::math::Pose3d> &_tfs,
+    ignition::math::Vector3d &_scale,
+    const std::string &_schemaToStop)
+  {
+    pxr::UsdPrim parent = _prim;
+    double metersPerUnit = 1.0;
+    std::string upAxis = "Y";
+
+    // this assumes that there can only be one stage
+    auto stageData = _usdData.FindStage(parent.GetPath().GetName());
+    if (stageData.second != nullptr) {
+      metersPerUnit = stageData.second->MetersPerUnit();
+      upAxis = stageData.second->UpAxis();
+    }
+
+    while (parent)
+    {
+      if (pxr::TfStringify(parent.GetPath()) == _schemaToStop)
+      {
+        return;
+      }
+
+      UDSTransforms t = ParseUSDTransform(parent);
+
+      ignition::math::Pose3d pose;
+      _scale *= t.scale;
+
+      pose.Pos() = t.translate * metersPerUnit;
+      // scaling is lost when we convert to pose, so we pre-scale the translation
+      // to make them match the scaled values.
+      if (!_tfs.empty()) {
+        auto& child = _tfs.back();
+        child.Pos().Set(
+          child.Pos().X() * t.scale[0],
+          child.Pos().Y() * t.scale[1],
+          child.Pos().Z() * t.scale[2]);
+      }
+
+      if (!t.isRotationZYX)
+      {
+        if (t.isRotation)
+        {
+          pose.Rot() = t.q[0];
```

Comment:
Assuming that is means rotationXYZ, do we need to decompose the order of rotation to 3 tfs (similar to ZYX case)?